synapse/events/builder.py

```diff
@@ -131,8 +138,14 @@ async def build(
             auth_events = auth_event_ids
             prev_events = prev_event_ids
 
-        old_depth = await self._store.get_max_depth_of(prev_event_ids)
-        depth = old_depth + 1
+        # Otherwise, progress the depth as normal
+        if depth is None:
```

Comment:
The last test failures are coming from this new line (if I comment it out, the tests pass). Is there a way to see `logger.info` from within the app while running the `twisted.trial` tests? I am having digging into this further without being able to log.

Test failure: https://github.com/matrix-org/synapse/pull/10049/checks?check_run_id=2764607442#step:8:2373

Reproduction locally:

```
python -m twisted.trial tests.handlers.test_presence.PresenceJoinTestCase.test_remote_gets_presence_when_local_user_joins
```

Reply:
Setting the env to `SYNAPSE_TEST_LOG_LEVEL=DEBUG` will give you debug logs in `_trial_temp/test.log` :slightly_smiling_face:

Reply:
It looks like I accidentally set `depth` to `False` in the tests.

Found the problem first by logging what the weird `depth` value was above the line in question and saw it was `False` at some point:
```python
logger.info("depth1=%s", depth)
```

Then I added this line which will cause it to throw an exception (because no `%s` in the formatting string)when this condition is met and give us a full stack trace.
```python
if depth == False:
    logging.info("Why is depth False?", depth)
```

```diff
SYNAPSE_TEST_LOG_LEVEL=DEBUG python -m twisted.trial tests.handlers.test_presence.PresenceJoinTestCase.test_remote_gets_presence_when_local_user_joins
tests.handlers.test_presence
  PresenceJoinTestCase
    test_remote_gets_presence_when_local_user_joins ...                  [FAIL]

===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/Users/eric/Documents/github/element/synapse/tests/handlers/test_presence.py", line 815, in test_remote_gets_presence_when_local_user_joins
    self._add_new_user(room_id, "@alice:server2")
-  File "/Users/eric/Documents/github/element/synapse/tests/handlers/test_presence.py", line 866, in _add_new_user
-    event = self.get_success(builder.build(prev_event_ids, None, False))
  File "/Users/eric/Documents/github/element/synapse/tests/unittest.py", line 500, in get_success
    return self.successResultOf(d)
  File "/Users/eric/.pyenv/versions/3.8.6/lib/python3.8/site-packages/twisted/trial/_synctest.py", line 706, in successResultOf
    self.fail(
twisted.trial.unittest.FailTest: Success result expected on <Deferred at 0x10de9a3a0 current result: None>, found failure result instead:
Traceback (most recent call last):
  File "/Users/eric/.pyenv/versions/3.8.6/lib/python3.8/site-packages/twisted/internet/defer.py", line 477, in callback
    self._startRunCallbacks(result)
  File "/Users/eric/.pyenv/versions/3.8.6/lib/python3.8/site-packages/twisted/internet/defer.py", line 580, in _startRunCallbacks
    self._runCallbacks()
  File "/Users/eric/.pyenv/versions/3.8.6/lib/python3.8/site-packages/twisted/internet/defer.py", line 662, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/Users/eric/.pyenv/versions/3.8.6/lib/python3.8/site-packages/twisted/internet/defer.py", line 1514, in gotResult
    current_context.run(_inlineCallbacks, r, g, status)
--- <exception caught here> ---
  File "/Users/eric/.pyenv/versions/3.8.6/lib/python3.8/site-packages/twisted/internet/defer.py", line 1445, in _inlineCallbacks
    result = current_context.run(g.send, result)
  File "/Users/eric/Documents/github/element/synapse/synapse/events/builder.py", line 142, in build
    logging.info("Why is depth False?", depth)
  File "/Users/eric/.pyenv/versions/3.8.6/lib/python3.8/logging/__init__.py", line 2070, in info
    root.info(msg, *args, **kwargs)
  File "/Users/eric/.pyenv/versions/3.8.6/lib/python3.8/logging/__init__.py", line 1434, in info
    self._log(INFO, msg, args, **kwargs)
  File "/Users/eric/.pyenv/versions/3.8.6/lib/python3.8/logging/__init__.py", line 1577, in _log
    self.handle(record)
  File "/Users/eric/.pyenv/versions/3.8.6/lib/python3.8/logging/__init__.py", line 1587, in handle
    self.callHandlers(record)
  File "/Users/eric/.pyenv/versions/3.8.6/lib/python3.8/logging/__init__.py", line 1649, in callHandlers
    hdlr.handle(record)
  File "/Users/eric/.pyenv/versions/3.8.6/lib/python3.8/logging/__init__.py", line 950, in handle
    self.emit(record)
  File "/Users/eric/Documents/github/element/synapse/tests/test_utils/logging_setup.py", line 28, in emit
    log_entry = self.format(record)
  File "/Users/eric/.pyenv/versions/3.8.6/lib/python3.8/logging/__init__.py", line 925, in format
    return fmt.format(record)
  File "/Users/eric/.pyenv/versions/3.8.6/lib/python3.8/logging/__init__.py", line 664, in format
    record.message = record.getMessage()
  File "/Users/eric/.pyenv/versions/3.8.6/lib/python3.8/logging/__init__.py", line 369, in getMessage
    msg = msg % self.args
builtins.TypeError: not all arguments converted during string formatting
```

---

Is there a better way to get a stack trace/call stack? I tried `logger.exception` and adding `exc_info=True` to the logger functions but they just gave small trace back to the `defer` and don't play well with the async stuff.

```
2021-06-09 01:56:40-0500 [-] 2021-06-09 01:56:40,722 - root - 142 - INFO - sentinel - Why is depth False?
	Traceback (most recent call last):
	  File "/Users/eric/.pyenv/versions/3.8.6/lib/python3.8/site-packages/twisted/internet/defer.py", line 1445, in _inlineCallbacks
	    result = current_context.run(g.send, result)
	StopIteration: _GetStateGroupDelta(prev_group=5, delta_ids={('m.room.history_visibility', ''): '$6KwfF_Su1zGCDsZn4FvqkfkcPqI2CnPn_-Uue7pXRCc'})
```

Reply:
Documented the `SYNAPSE_TEST_LOG_LEVEL=DEBUG`( -> `_trial_temp/test.log`) trick in https://github.com/matrix-org/synapse/pull/10148. Thanks for sharing!

Reply:
Can you raise if depth is wrong? That gives me a good stack trace. I think the defer code does special things to make the stack traces work out?

Reply:
`raise` does work well :+1: Thanks for the better way to do this!

Is there a way I can just log a line and get a stack trace without breaking the normal execution flow?

In JavaScript, I would use `console.log('something bad happened', new Error().stack);`

Reply:
I *think* creating a twisted `Failure` object might work: https://github.com/matrix-org/synapse/blob/develop/synapse/handlers/pagination.py#L285-L288, but I'm not 100% sure